examples/exception_handling/exception_handling.py

```diff
@@ -0,0 +1,90 @@
+import asyncio
+from typing import Callable, Tuple
+
+from golem_core import GolemNode, Payload, commands
+from golem_core.default_logger import DefaultLogger
+from golem_core.default_payment_manager import DefaultPaymentManager
+from golem_core.low.activity import Activity
+from golem_core.low.exceptions import BatchError
+from golem_core.mid import (
+    Buffer,
+    Chain,
+    Limit,
+    Map,
+    default_create_activity,
+    default_create_agreement,
+    default_negotiate,
+)
+
+PAYLOAD = Payload.from_image_hash(
+    "9a3b5d67b0b27746283cb5f287c13eab1beaa12d92a9f536b747c7ae"
+)
+SUBNET = "public"
+
+
+async def trigger_exception(activity: Activity) -> str:
+    provider_id = activity.parent.parent.data.issuer_id
+    try:
+        batch = await activity.execute_commands(
+            commands.Deploy(),
+            commands.Start(),
+            commands.Run(["/golem/entrypoints/run-blender.sh"]),
+        )
+    except BatchError:
+        # Handle exceptions directly in handler func
+        raise
+    await batch.wait(timeout=20)
+
+    return f"{provider_id}"
+
+
+async def on_exception(func: Callable, args: Tuple, e: Exception) -> None:
+    activity = args[0]
+    print(f"Activity {activity} failed because of:\n{e}")
+    # Handle exception in `on_exception` func
+    if isinstance(e, BatchError):
+        print(f"Batch stdout: {e.batch.events[-1].stdout}")
+        print(f"Batch stderr: {e.batch.events[-1].stderr}")
+    await activity.parent.close_all()
+
+
+async def main() -> None:
+    golem = GolemNode()
+    golem.event_bus.listen(DefaultLogger().on_event)
+
+    async with golem:
+        allocation = await golem.create_allocation(1.0)
+        payment_manager = DefaultPaymentManager(golem, allocation)
+        demand = await golem.create_demand(
+            PAYLOAD, allocations=[allocation], subnet=SUBNET
+        )
+
+        chain = Chain(
+            demand.initial_proposals(),
+            Map(default_negotiate),
+            Map(default_create_agreement),
+            Map(default_create_activity),
+            Limit(1),
+            Map(trigger_exception, on_exception=on_exception),  # type: ignore[arg-type]
+            Buffer(1),
+        )
+
+        async for result in chain:
+            print(f"Finished with {result}")
+
+        print("TASK DONE")
+        await payment_manager.terminate_agreements()
+        await payment_manager.wait_for_invoices()
+
+
+if __name__ == "__main__":
+    loop = asyncio.get_event_loop()
+    task = loop.create_task(main())
+    try:
+        loop.run_until_complete(task)
+    except KeyboardInterrupt:
+        task.cancel()
+        try:
+            loop.run_until_complete(task)
+        except asyncio.CancelledError:
+            pass
```

Comment:
This whole block is handled by default in [`asyncio.run`](https://github.com/python/cpython/blob/3.10/Lib/asyncio/runners.py#L47), even with cancellation attempt.

```suggestion
    asyncio.run(main())
```

This pattern should be even propagated everywhere in our code.

Reply:
ctr-c didn't work for me when I used `asyncio.run(main())`. I had to send it twice for some reason.